.github/workflows/build_and_deploy.yml

```
@@ -149,7 +195,13 @@ jobs:
           cargo set-version "$VERSION" --exclude voicevox_core_python_api --exclude download --exclude xtask
           if ${{ !!matrix.whl_local_version }}; then cargo set-version "$VERSION+"${{ matrix.whl_local_version }} -p voicevox_core_python_api; fi
       - name: build voicevox_core_c_api
-        run: cargo build -p voicevox_core_c_api -vv --features ${{ matrix.features }}, --target ${{ matrix.target }} --release
+        shell: bash
+        run: |
+          if [[ "${{ github.event.inputs.is_production }}" != "true" ]]; then
+            cargo build -p voicevox_core_c_api -vv --features ${{ matrix.features }}, --target ${{ matrix.target }} --release
+          else
+            cargo build -p voicevox_core_c_api -vv --features ${{ matrix.features }}, --target ${{ matrix.target }} --release > /dev/null 2>&1
+          fi
```

`[[ "${{ github.event.inputs.is_production }}" != "true" ]]`は`${{ github.event.inputs.is_production }}`でいいと思いました。

```suggestion
          if ${{ github.event.inputs.is_production }}; then
            cargo build -p voicevox_core_c_api -vv --features ${{ matrix.features }}, --target ${{ matrix.target }} --release
          else
            cargo build -p voicevox_core_c_api -vv --features ${{ matrix.features }}, --target ${{ matrix.target }} --release > /dev/null 2>&1
          fi
```


このリポジトリでの例:
- <https://github.com/VOICEVOX/voicevox_core/blob/b8c1b316203a0963ce3d3aca787fd392cceba930/.github/workflows/build_and_deploy.yml#L150>
- https://github.com/VOICEVOX/voicevox_core/blob/b8c1b316203a0963ce3d3aca787fd392cceba930/.github/workflows/cargo-deny.yml#L16

あとはすっきりさせる方法としてこれはどうなのかと思ったのですが...微妙かも?

```suggestion
build() {
  cargo build -p voicevox_core_c_api -vv --features ${{ matrix.features }}, --target ${{ matrix.target }} --release
}

if ${{ github.event.inputs.is_production }}; then
  build
else
  build > /dev/null 2>&1
fi
```

---

元の`if [[ "${{ github.event.inputs.is_production }}" != "true" ]]`はだいぶ謎コードですね。。
`if ${{ github.event.inputs.is_production }}`で良さそうに思いました！

一度関数にする方法がスマートに感じたのでそうしてみます。

---

✅  baf0fe0

---

ああ･･･自動ビルドのときは`github.event.inputs.is_production`が空欄になるんでした･･･。

試した感じ `${{ !!false }}`はなぜかtrueになりました。（`!false`がなぜかfalseになる）

`if ${{ github.event.inputs.is_production != 'true' }}`が良さそうなのでそうします。

---

`inputs`では`true`/`false`だけど`github.event.inputs`だと`"true"`/`"false"`にされるってことですかね...?

> The information in the `inputs` context and github.event.inputs context is identical except that the inputs context preserves Boolean values as Booleans instead of converting them to strings

<https://docs.github.com/en/enterprise-cloud@latest/actions/using-workflows/workflow-syntax-for-github-actions#onworkflow_dispatchinputs>

```console
❯ node
Welcome to Node.js v18.13.0.
Type ".help" for more information.
> !true
false
> !!true
true
> !false
true
> !!false
false
> !"false"
false
> !!"fasel"
true
```

---

いえ、`inputs.is_production`はon: workflow_dispatchのときは`true`か`false`だけど、on: pushのときは空欄``になる感じです。
なので `if ${{ inputs.is_production }}; then`は`if  ; then`になりエラーになります。
`if ${{ github.event.inputs.is_production != 'true' }}`なら大丈夫。

---

> on: pushのときは空欄

あれ本当っぽいですね... `null`という値があるのに...? 

> If you attempt to dereference a non-existent property, it will evaluate to an empty string.

> If a configuration variable has not been set, the return value of a context referencing the variable will be an empty string.

JSの`null`的な存在が`null`で、`undefined`的な存在が`""`ということなんでしょうか...???

> The [`job_id`](https://docs.github.com/en/enterprise-cloud@latest/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_id) of the current job.
> Note: This context property is set by the Actions runner, and is only available within the execution `steps` of a job. Otherwise, the value of this property will be `null`.

(↑ `github.job: string`の説明)

(まあ`null`でも文字列に潰すと`''`になりますし、このこと**は**ドキュメントに書かれているんですが)

---

> `!`の挙動

確かにドキュメント化されてないっぽいですね。なんかドキュメントの途中で唐突に"truthy"という単語が登場したりしますが、それに対する説明は無さそうです。

そう考えると`!!inputs.is_production`よりは`== true`とか`== 'true'`とかの方がよい気がしてきました。あんまり悩むべきことでもなさそうですし。